### DISCUSSION TITLE
Behavior of fflush(NULL) in Picolibc's TinyStdio

### DISCUSSION THREAD

**Discussion by M-Moawad** [picolibc/picolibc#723] [Q&A] [answered]
Hello everyone,

I'd like to start a discussion about the behavior of `fflush(NULL)` in Picolibc's tinystdio. After investigating and testing this functionality, I found that the current implementation throws an exception due to accessing a null pointer.

**Expected Behavior:** According **ISO_IEC_9899_1999 section 7.19.5.2**: If the stream is a null pointer, the `fflush()` performs this flushing action on all streams.

**Note:** Picolibc's stdio version, this feature is implemented by adding a data structure that tracks the opened buffers.
Here: https://github.com/picolibc/picolibc/blob/main/newlib/libc/stdio/stdio.h
```c
struct _glue
{
  struct _glue *_next;
  int _niobs;
  __FILE *_iobs;
};
```
Elements are added to this data structure when a file is opened using `fopen()` and removed when `fclose()` is called.

One potential approach could involve adapting the data structure used in the standard **stdio** version to track opened buffers and implementing it within the **tinystdio**. However, this requires careful consideration and planning to ensure seamless integration without compromising tinystdio's lightweight and efficient design principles.

So what are your insights and suggestions on how best to implement this enhancement in Picolibc's tinystdio?

Looking forward to hearing your thoughts and ideas on this matter.

Best regards,
Mohamed Moawad

**Comment by keith-packard** [comment]
Hrm. A couple of thoughts:

 * Not all FILEs have an fflush hook, those without wouldn't need to participate here.
 * This only matters if the app calls fflush(NULL)
 * We need to deal with stdout/stderr too
 * Do we care about other FILE structs which the application may allocate itself?

So, I'd propose that we add a function in the same file as fflush which can be called when allocating a FILE with an fflush hook. Declare that `weak` and only call if non-NULL so that when fflush is not used by the app, the overhead will disappear. Rename the existing `fflush` as `_fflush` and use that whenever flushing is needed within the library with an explicit non-NULL FILE. 

Maybe even make `fflush` an `extern inline` function which tests its argument for NULL before calling the fancy code? We can annotate other stdio APIs with `attribute((nonnull))` for their FILE arguments so the compiler can bypass the fancy code when building applications. Some experimentation here would be helpful.

Use `weak` attributes on stdout/stderr in the fflush implementation to avoid pulling those in while still flushing them when they are used.

**Comment by keith-packard** [comment]
Well, that didn't go as I had hoped -- `fflush(stdout)` is pretty common, and there's no way for the compiler to know if `stdout` is null, so every one of those calls ends up using the full version.

However, would it be reasonable to suppose that users looking to flush all active files will use a constant expression with value NULL rather than a runtime computation which evaluates to NULL? If so, then using `__builtin_constant_p` and only flushing everything if the expression was both constant *and* NULL would avoid pulling in the full version in most cases.

**Comment by keith-packard** [comment, marked as answer]
Checking for a NULL constant seems to work fine; checkout PR #726 and see what you think. If only supporting constant NULL doesn't seem reasonable, then I think we punt on this bug.

**Comment by keith-packard** [comment]
Another simpler alternative is to have fflush(NULL) flush just stdout and stderr. That might be a less surprising semantic than having fflush(NULL) only "work" when the parameter is the constant NULL and not when it's being passed in from another function?
